src/java/com/puppetlabs/puppetserver/pool/JRubyPool.java

```diff
@@ -1,234 +1,371 @@
+    // Lock which guards all accesses to the underlying queue and registered
+    // element set.  Constructed as "nonfair" for performance, like the lock
+    // that a `LinkedBlockingDeque` does.  Not clear that we need this
+    // to be a "fair" lock.
+    private final ReentrantLock lock = new ReentrantLock();
+
+    // Condition signaled when all elements that have been registered have been
+    // returned to the queue.  Awaited when a lock has been requested but
+    // one or more registered elements has been borrowed from the pool.
+    private final Condition allRegisteredInQueue = lock.newCondition();
+
+    // Condition signaled when an element has been added into the queue.
+    // Awaited when a request has been made to borrow an item but no elements
+    // currently exist in the queue.
+    private final Condition notEmpty = lock.newCondition();
+
+    // Condition signaled when the pool has been unlocked.  Awaited when a
+    // request has been made to borrow an item or lock the pool but the pool
+    // is currently locked.
+    private final Condition notWriteLocked = lock.newCondition();
+
+    // Holds a reference to all of the elements that have been registered.
+    // Newly registered elements are also added into the `liveQueue`.
+    // Elements only exist in the `liveQueue` when not currently
+    // borrowed whereas elements that have been registered (but not
+    // yet unregistered) will be accessible via `registeredElements`
+    // even while they are borrowed.
+    private final Set<E> registeredElements = new CopyOnWriteArraySet<>();
+
+    // Maximum size that the underlying queue can grow to.
+    private int maxSize;
 
+    // Thread which currently holds the write lock.  null indicates that
+    // there is no current write lock holder.  Using the current Thread
+    // object for tracking the lock owner is comparable to what the JDK's
+    // `ReentrantLock` class does:
+    //
+    // http://hg.openjdk.java.net/jdk8/jdk8/jdk/file/687fd7c7986d/src/share/classes/java/util/concurrent/locks/ReentrantLock.java#l164
+    private Thread writeLockThread = null;
 
     /**
-     * This method removes an element from the list of "registered" elements,
-     * such that it will no longer be returned by calls to
-     * <tt>getRegisteredInstances</tt>.
-     *
-     * This method does not remove the element from the underlying queue; it
-     * is assumed that the caller has already done so via the methods of the
-     * parent class.
+     * Create a JRubyPool
      *
-     * @param e the element to remove from the list of registered instances.
+     * @param size maximum capacity for the pool.
      */
+    public JRubyPool(int size) {
+        liveQueue = new LinkedList<>();
+        maxSize = size;
+    }
+
     @Override
-    synchronized public void unregister(E e) throws InterruptedException {
-        registeredElements.remove(e);
+    public void register(E e) {
+        final ReentrantLock lock = this.lock;
```

Comment:
Yeah, just trying to be idiomatic per how the Java concurrency class implementations are written.  There may be some subtle benefits to this that I don't understand but this didn't seem harmful in any event.